de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/negotiation/InstantIncomingProjectNegotiation.java

```diff
@@ -0,0 +1,127 @@
+import de.fu_berlin.inf.dpp.filesystem.IWorkspace;
+import de.fu_berlin.inf.dpp.monitoring.IProgressMonitor;
+import de.fu_berlin.inf.dpp.negotiation.NegotiationTools.CancelOption;
+import de.fu_berlin.inf.dpp.negotiation.stream.IncomingStreamProtocol;
+import de.fu_berlin.inf.dpp.net.IReceiver;
+import de.fu_berlin.inf.dpp.net.ITransmitter;
+import de.fu_berlin.inf.dpp.net.xmpp.JID;
+import de.fu_berlin.inf.dpp.net.xmpp.XMPPConnectionService;
+import de.fu_berlin.inf.dpp.observables.FileReplacementInProgressObservable;
+import de.fu_berlin.inf.dpp.session.ISarosSession;
+import de.fu_berlin.inf.dpp.session.ISarosSessionManager;
+
+/**
+ * Receive shared Projects and display them instant using a stream based
+ * solution.
+ */
+public class InstantIncomingProjectNegotiation extends
+    AbstractIncomingProjectNegotiation {
+
+    private static final Logger log = Logger
+        .getLogger(InstantIncomingProjectNegotiation.class);
+
+    public InstantIncomingProjectNegotiation(
+        final JID peer, //
+        final String negotiationID, //
+        final List<ProjectNegotiationData> projectNegotiationData, //
+        final ISarosSessionManager sessionManager, //
+        final ISarosSession session, //
+        final FileReplacementInProgressObservable fileReplacementInProgressObservable, //
+        final IWorkspace workspace, //
+        final IChecksumCache checksumCache, //
+        final XMPPConnectionService connectionService, //
+        final ITransmitter transmitter, //
+        final IReceiver receiver //
+    ) {
+        super(peer, TransferType.INSTANT, negotiationID,
+            projectNegotiationData, sessionManager, session,
+            fileReplacementInProgressObservable, workspace, checksumCache,
+            connectionService, transmitter, receiver);
+    }
+
+    @Override
+    protected void transfer(IProgressMonitor monitor,
+        Map<String, IProject> projectMapping, List<FileList> missingFiles)
+        throws IOException, SarosCancellationException {
+
+        awaitActivityQueueingActivation(monitor);
+
+        /*
+         * the user who sends this ProjectNegotiation is now responsible for the
+         * resources of the contained projects
+         */
+        for (Entry<String, IProject> entry : projectMapping.entrySet()) {
+            final String projectID = entry.getKey();
+            final IProject project = entry.getValue();
+
+            session.addProjectMapping(projectID, project);
+            /* TODO change queuing to resource based queuing */
+            session.enableQueuing(project);
+        }
+
+        transmitter.send(ISarosSession.SESSION_CONNECTION_ID, getPeer(), //
+            StartActivityQueuingResponse.PROVIDER //
+                .create( //
+                new StartActivityQueuingResponse(getSessionID(), getID())));
+
+        checkCancellation(CancelOption.NOTIFY_PEER);
+
+        /* only get files, if something is missing */
+        int filesMissing = 0;
```

Comment:
You could fix the naming thing (using `filesMissing` and `missingFiles` for completely different purposes) in this PR or in #156. It does not really matter to me.